### PR TITLE
Check for local service before deletion

### PR DESCRIPTION
**Pull request by ajssmith**
Check that a service has not been exposed locally prior to deletion.

### CODE REVIEW

File: cmd/service-controller/service_sync.go [hunk 1]
```diff
@@ -118,10 +106,13 @@ func (c *Controller) ensureServiceInterfaceDefinitions(origin string, serviceInt
 		current := c.byOrigin[origin]
 		for name, _ := range current {
 			if _, ok := serviceInterfaceDefs[name]; !ok {
-				deleted = append(deleted, name)
+				if _, local := c.localServices[name]; !local {
```
[the remark this answers is not in the document]
Comment (grs) [reply cmd/service-controller/service_sync.go:109]: Is the root of the bug perhaps that we don't ever delete from byOrigin, even if the origin in the local definition changes?

[the remark this answers is not in the document]
Comment (ajssmith) [reply cmd/service-controller/service_sync.go:109]: Yes, that would be the root issue. I can see a better way to address the issue and will update the pr. thnx.

File: cmd/service-controller/service_sync.go [hunk 2]
```diff
@@ -36,33 +44,27 @@ func (c *Controller) serviceSyncDefinitionsUpdated(definitions map[string]types.
 			if _, ok := c.byOrigin[service.Origin]; !ok {
 				c.byOrigin[service.Origin] = make(map[string]types.ServiceInterface)
 			}
-			c.byOrigin[service.Origin][name] = service
+			// track by origin if not locally exposed
+			if _, ok := c.localServices[name]; !ok {
```
Comment (grs) [inline cmd/service-controller/service_sync.go:48]: I don't think this check should be here. The `definitions` map holds the latest local service definitions (i.e. the services as defined in the configmap). The `localServices` map holds the previous set of service definitions.

E.g. in a scenario where the local service is unexposed and immediately after we get a service sync update from another site containing that service, the definition in the configmap would be updated to have the new origin. However the localServices will still hold the stale definition.

I believe the origin in the records in  `definitions` is authoritative here. (If we wanted we could ensure that the definition can never have both an originat that is non-null and not "annotations" *and* a target defined. The presence of a target would invalidate an explicit origin.

Comment (ajssmith) [reply cmd/service-controller/service_sync.go:48]: Good catch. Will remove check and do additional testing including annotations, full mesh, etc.

**Pull request merged**